lte/gateway/c/session_manager/SessionState.cpp

```diff
@@ -2784,4 +2784,212 @@ bool operator==(const Teids& lhs, const Teids& rhs) {
+  return to_process;
+}
+
+RuleToProcess SessionState::insert_dynamic_5g_rule(
+    const PolicyRule& rule, const RuleLifetime& lifetime,
+    SessionStateUpdateCriteria* session_uc) {
+  RuleToProcess to_process;
+  rule_lifetimes_[rule.id()] = lifetime;
+  dynamic_rules_.insert_rule(rule);
+  if (session_uc) {
+    session_uc->dynamic_rules_to_install.push_back(rule);
+    session_uc->new_rule_lifetimes[rule.id()] = lifetime;
+  }
+  increment_rule_stats(rule.id(), session_uc);
+
+  to_process.version = get_current_rule_version(rule.id());
+  to_process.rule    = rule;
+  to_process.teids   = config_.common_context.teids();
+
+  return to_process;
+}
+
+optional<RuleToProcess> SessionState::remove_dynamic_5g_rule(
+    const std::string& rule_id, PolicyRule* rule_out,
+    SessionStateUpdateCriteria* session_uc) {
+  PolicyRule rule;
+  RuleToProcess to_process;
+  bool removed = dynamic_rules_.remove_rule(rule_id, &rule);
+  if (!removed) {
+    return {};
+  }
+  if (rule_out) {
+    *rule_out = rule;
+  }
+  if (session_uc) {
+    session_uc->dynamic_rules_to_uninstall.insert(rule_id);
+  }
+  increment_rule_stats(rule_id, session_uc);
+
+  to_process.version = get_current_rule_version(rule.id());
+  to_process.rule    = rule;
+  to_process.teids   = config_.common_context.teids();
+
+  return to_process;
+}
+
+void SessionState::process_get_5g_rule_installs(
+    const std::vector<StaticRuleInstall>& static_rule_installs,
+    const std::vector<DynamicRuleInstall>& dynamic_rule_installs,
+    RulesToProcess* pending_activation, RulesToProcess* pending_deactivation) {
+  for (const StaticRuleInstall& rule_install : static_rule_installs) {
+    const std::string& rule_id = rule_install.rule_id();
+    PolicyRule rule;
+    RuleToProcess to_process;
+    static_rules_.get_rule(rule_id, &rule);
+    to_process.version = get_current_rule_version(rule.id());
+    to_process.rule    = rule;
+    to_process.teids   = config_.common_context.teids();
+    pending_activation->push_back(to_process);
+    pending_deactivation->push_back(to_process);
+  }
+  for (const DynamicRuleInstall& rule_install : dynamic_rule_installs) {
+    const PolicyRule& dynamic_rule = rule_install.policy_rule();
+    RuleToProcess to_process;
+    to_process.version = get_current_rule_version(dynamic_rule.id());
+    to_process.rule    = dynamic_rule;
+    to_process.teids   = config_.common_context.teids();
+    pending_activation->push_back(to_process);
+    pending_deactivation->push_back(to_process);
+  }
+}
+
+void SessionState::remove_all_5g_rules_for_termination(
+    SessionStateUpdateCriteria* session_uc) {
+  std::vector<PolicyRule> dynamic_rules;
+  dynamic_rules_.get_rules(dynamic_rules);
+  for (PolicyRule& policy : dynamic_rules) {
+    remove_dynamic_5g_rule(policy.id(), nullptr, session_uc);
+  }
+  for (const std::string& rule_id : active_static_rules_) {
+    deactivate_static_5g_rule(rule_id, session_uc);
+  }
+}
+
```

Comment:
Many of those functions are copy paste of 4g path.  Why are we copy&pasting instead of reusing?

Reply:
Same as above answer.

Reply:
Now, we are reusing the code.